packages/react/src/utils/useScrollLock.ts

```diff
@@ -137,66 +138,94 @@ function preventScrollStandard(referenceElement: Element | null) {
   };
 }
 
+class ScrollLocker {
+  lockCount = 0;
+
+  restore = null as (() => void) | null;
+
+  timeoutLock = Timeout.create();
+  timeoutUnlock = Timeout.create();
+
+  acquire(referenceElement: Element | null) {
+    this.lockCount += 1;
+    if (this.lockCount === 1 && this.restore === null) {
+      this.timeoutLock.start(0, () => this.lock(referenceElement));
+    }
+    return this.release;
+  }
+
+  release = () => {
+    this.lockCount -= 1;
+    this.timeoutUnlock.start(0, this.unlock);
+  };
+
+  private unlock = () => {
+    if (this.lockCount === 0 && this.restore) {
+      this.restore?.();
+      this.restore = null;
+    }
+  };
```

Comment:
The unlock timeout avoids unlock-lock sequences that create a lot of work.

![Screenshot From 2025-05-14 04-16-43](https://github.com/user-attachments/assets/3f21290d-ef79-4b8f-87c4-8115f5b231b5)